### PR TITLE
implement improved xDS NACK semantics

### PR DESCRIPTION
**Pull request by markdroth**
As per grpc/proposal#260.

### CODE REVIEW

File: test/cpp/end2end/xds_end2end_test.cc [hunk 1]
```diff
@@ -7255,12 +7332,54 @@ TEST_P(CdsTest, MultipleBadResources) {
...
+      {"cluster", kDefaultClusterName},
+  };
+  CheckRpcSendOk(
+      1, RpcOptions().set_metadata(std::move(metadata_default_cluster)));
+  // RPCs for cluster 2 should fail.
+  std::vector<std::pair<std::string, std::string>> metadata_cluster_2 = {
+      {"cluster", kClusterName2},
+  };
+  CheckRpcSendFailure(CheckRpcSendFailureOptions().set_rpc_options(
+      RpcOptions().set_metadata(std::move(metadata_default_cluster))));
```
Comment (donnadionne) [inline test/cpp/end2end/xds_end2end_test.cc:7352]: metadata_default_cluster was moved a second time

**Review by donnadionne — APPROVED** [review]
just a minor comment

File: test/cpp/end2end/xds_end2end_test.cc (hunk 1 above)
Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc]: Thanks for catching this!  That's actually a bug in the test; it was supposed to use the second set of metadata, not reuse the first set.

Fixed.

**Comment by markdroth** [comment]
The "Artifact Build MacOS" failure looks like an infrastructure timeout.

**Pull request merged**